### ISSUE TITLE
Slow startup time of app initialization and logic

### ISSUE THREAD

**Issue by z0xrv** [florisboard/florisboard#1650]
#### Short description
If i take a long time not using the keyboard (~30min) and if I tried to write something the smart bar will appear and after 3 seconds or so the elemnts of keyboard will apear(letters).
Also, when I click the emoji button it lags for 3 or 4 seconds also.
Note: This often happened but when I used ignoring battery optimization and helped reduce how often this happens

#### Steps to reproduce
1. Do not use the keyboard for some time
2. Try to use the keyboard
3. See the bug

#### Environment information
- FlorisBoard Version: 0.3.14-rc01
- Install Source: F-Droid
- Device: Condor P6 pro
- Android: 7.0
``

**Comment by lawson85** [comment]
Sounds like the OS is killing it, follow steps at dontkillmyapp.com to prevent the OS from killing it

**Comment by z0xrv** [comment]
But this happens only with florisboard not the other keyboards

**Comment by patrickgold** [comment]
> If i take a long time not using the keyboard (~30min) and if I tried to write something the smart bar will appear and after 3 seconds or so the elemnts of keyboard will apear(letters).

That definitely sounds like the OS is killing FlorisBoard :(

I think this has something to do with the fact that my keyboard has some optimization issues in terms of idling, thus the battery detection logic thinks I drain the battery too much and just outright kills my keyboard. Not too sure how I can fix this, I think fixing this is a long-term effort and involves a lot of debugging. Leaving open for now because I rarely notice this too, although for me it is far less frequent.

**Comment by tsiflimagas** [comment]
I speculate that there are probably two problems exhibited here. FlorisBoard frequently being killed by the system is one, perhaps (at least I think I've not experienced it). But the other one is that startup time is indeed a bit long. I notice it on my phone too (not as vividly). It takes like 0.5-1s for the keyboard to be fully drawn, if it's cold started (otherwise it's instant). It may not seem much to the eye, but when compared to other keyboards, it makes you suspicious. Also, interaction seems to be a little slow, for a short period, initially. Although input seems to be moving normally fast (after the work done on 0.3.16-beta01, UI seems to be lagging behind a bit. I think that's amplified for the issue author, because his cpu is rather weak, let's say, from what I looked up (not saying anything in bad will, of course, just investigating).
Since UI was migrated to Jetpack Compose, some performance issues occured, in some parts of it. What concerns me, is that I noticed the same on some others projects too, since they switched to the same library, mostly laggy scrolling and the (in question) slow startup. I wonder if there's actually some issue in the library itself and whether it's known and tracked, if that's the case. I'm not a dev, so my speculations could be totally off.
I read something about baseline profiles, which are suggested as a solution to improve startup performance. Perhaps it's worthy looking into it(?)

**Comment by patrickgold** [comment]
> Another common point is this kind of message in logcat
W orisboard.debu: Class androidx.compose.runtime.snapshots.SnapshotStateList failed lock verification

I actually noticed this too and I think either you or someone else mentioned this while still in the 0.3.14 beta phase. The biggest problem for me is that there's not much info online about this error (specifically in conjunction with JC). However while I think this is an issue, this would cause all `SnapshotStateList`s to run slower not only during startup but also just while using the keyboard. So I don't think this specifically plays too much of a factor here when just looking at the startup time.

Generally though the JC library is (sadly) known for having a not so great startup time even for empty hello world projects, so the library may be parts of the slow startup time.

> I read something about baseline profiles, which are suggested as a solution to improve startup performance. Perhaps it's worthy looking into it(?)

Thanks for bringing this topic up, this is something new to me! I looked up the baseline profiles and there seems to be a real performance increase possible when generating the baseline profile correctly. This is something which I will work on getting implemented after finishing and merging #1860 in (provided the baseline profile really makes an impact).

**Comment by tsiflimagas** [comment]
Well, the logcat message is now gone and, though I hope this is not placebo (I think it's not), I can see the app starting up faster and being responsive right away. Nice job! It seems to me that performance got a little boost in general. I'm interested to hear what others will notice too after the next beta release. 

> Generally though the JC library is (sadly) known for having a not so great startup time even for empty hello world projects, so the library may be parts of the slow startup time.

Ah hell, so it does have some performance issues. Well, at least the burden is not on you, but on the team working on it, and since it's a big project, I suppose they work on it intensively.

> Thanks for bringing this topic up, this is something new to me! I looked up the baseline profiles and there seems to be a real performance increase possible when generating the baseline profile correctly.

Oh, good to see that it can be useful! I bumped into it while searching about JC and performance. What made me some impression while at it, was that it seems fairly easy to run into bad performance through wrong practices (i watched a video by Google with some examples). Or perhaps that's software in general, so nothing special 😄

**Comment by patrickgold** [comment]
> Well, the logcat message is now gone and, though I hope this is not placebo (I think it's not), I can see the app starting up faster and being responsive right away. Nice job! It seems to me that performance got a little boost in general. I'm interested to hear what others will notice too after the next beta release.

That's true, this is something I didn't even notice until you said it. I think it could be because with #1860 I introduced Proguard optimizations, which strips unused code. Maybe the error comes from one part of the JC library which I don't use and thus the source of this issue is now gone.

As for performance increase at startup: The beta and release builds I tested seem a bit more responsive and with #1870 there's a slight improvement again for the startup time. It will be really interesting how the general feedback for this performance improvement is in the next beta release.

Still, the startup time is far from perfect and while testing earlier today I also know why: For every cold start of the app/keyboard, I do very heavy disk read operations (indexing of extensions, mapping layouts, loading extension meta data), only after all that is done the layout manager can read the active subtype's layout file, parse it and pass it to the UI layer for drawing. The Smartbar does not have to do this, thus the Smartbar appears way quicker than the actual keyboard. This means I have to go full into debugging and improving the way extension indexing works.

> Ah hell, so it does have some performance issues. Well, at least the burden is not on you, but on the team working on it, and since it's a big project, I suppose they work on it intensively.

That's for the general startup time, yeah. But as mentioned above, I also have to do some fixing, not every ms delayed is caused by JC :)

> What made me some impression while at it, was that it seems fairly easy to run into bad performance through wrong practices (i watched a video by Google with some examples). Or perhaps that's software in general, so nothing special

Oh yeah, that's soo true! However that's absolutely not specific to Android development, but software in general :)

**Comment by lm41** [comment]
This issue has been open for a long period of time. Does it still occur in the latest version?
